bip-0346.md

```diff
@@ -0,0 +1,438 @@
+| `NONE`                       | `0b11110011`           |
+| `ALL|ANYONECANPAY`           | `0b11111101`           |
+| `SINGLE|ANYONECANPAY`        | `0b11110101`           |
+| `NONE|ANYONECANPAY`          | `0b11110001`           |
+| `ALL|ANYPREVOUT`             | `0b11101101`           |
+| `SINGLE|ANYPREVOUT`          | `0b11100101`           |
+| `NONE|ANYPREVOUT`            | `0b11100001`           |
+| `ALL|ANYPREVOUTANYSCRIPT`    | `0b11001101`           |
+| `SINGLE|ANYPREVOUTANYSCRIPT` | `0b11000101`           |
+| `NONE|ANYPREVOUTANYSCRIPT`   | `0b11000001`           |
+
+* The first byte of the TxFieldSelector has its 8 bits assigned as follows, from lowest to highest:
+  * 1: version (`TXFS_VERSION`)
+  * 2: locktime (`TXFS_LOCKTIME`)
+  * 3: current input index (`TXFS_CURRENT_INPUT_IDX`)
+  * 4: current input control block (or empty) (`TXFS_CURRENT_INPUT_CONTROL_BLOCK`)
+  * 5: current input spent script (i.e. witness script or tapscript) (`TXFS_CURRENT_INPUT_SPENTSCRIPT`)
+  * 6: current script last `OP_CODESEPARATOR` position (or 0xffffffff)
+    (`TXFS_CURRENT_INPUT_LAST_CODESEPARATOR_POS`)
+  * 7: current input annex (or empty) (`TXFS_CURRENT_INPUT_TAPROOT_ANNEX`)
+  * 8: `TXFS_CONTROL` (i.e. include TxFieldSelector into hash)
+
+* The highest bit of the first byte (`TXFS_CONTROL`), we will call the
+  "control bit", and it can be used to control the behavior of the opcode. For
+  `OP_TXHASH` and `OP_CHECKTXHASHVERIFY`, the control bit is used to determine
+  whether the TxFieldSelector itself has to be included in the resulting hash.
+  (For potential other uses of the TxFieldSelector (like a hypothetical
+  `OP_TX`), this bit can be repurposed.)
+
+* The second byte will be used to indicate fields from the inputs and outputs.
+  If there is only a single byte present, no information from the inputs and
+  outputs will be committed. Otherwise, of the second byte, the 8 bits are
+  assigned the following variables, from lowest to highest:
+  * Specifying which fields of the inputs will be selected:
+    * 1: prevouts (`TXFS_INPUTS_PREVOUTS`)
+    * 2: sequences (`TXFS_INPUTS_SEQUENCES`)
+    * 3: scriptSigs (`TXFS_INPUTS_SCRIPTSIGS`)
+    * 4: prevout scriptPubkeys (`TXFS_INPUTS_PREV_SCRIPTPUBKEYS`)
+    * 5: prevout values (`TXFS_INPUTS_PREV_VALUES`)
+    * 6: taproot annexes (`TXFS_INPUTS_TAPROOT_ANNEXES`)
+
+  * Specifying which fields of the outputs will be selected:
+    * 7: scriptPubkeys (`TXFS_OUTPUTS_SCRIPTPUBKEYS`)
+    * 8: values (`TXFS_OUTPUTS_VALUES`)
+
+* We define as follows:
+  * `TXFS_ALL = TXFS_VERSION | TXFS_LOCKTIME | TXFS_CURRENT_INPUT_IDX | TXFS_CURRENT_INPUT_CONTROL_BLOCK | TXFS_CURRENT_INPUT_LAST_CODESEPARATOR_POS | TXFS_CONTROL`
+  * `TXFS_INPUTS_ALL = TXFS_INPUTS_PREVOUTS | TXFS_INPUTS_SEQUENCES | TXFS_INPUTS_SCRIPTSIGS | TXFS_INPUTS_PREV_SCRIPTPUBKEYS | TXFS_INPUTS_PREV_VALUES | TXFS_INPUTS_TAPROOT_ANNEXES`
+  * `TXFS_OUTPUTS_ALL = TXFS_OUTPUTS_SCRIPTPUBKEYS | TXFS_OUTPUTS_VALUES`
```

Comment:
Please explain the semantics of these lines. Does the pipe represent string concatenation, or some other form of concatenation here?